### ISSUE TITLE
mixin implements in nested

### ISSUE THREAD

**Issue by zmtzawqlp** [dart-lang/sdk#50167]
class D must override test and test1 methods?  A has already done this， is this an issue?

``` dart
class F {}

mixin B {
  void test() {}
}

mixin A on F implements B {
  @override
  void test() {
    print('ddd');
  }

  void test1() {}
}

mixin C on F implements A {}

class D extends F with C {
  @override
  void test() {
    // TODO: implement test
  }

  @override
  void test1() {
    // TODO: implement test1
  }
}
```

**Comment by lrhn** [comment]
The `D` class only `implements A`, it does not extend `A` or mix it in, so it gets no method *implementation* from `A`.

Since the super class `F` has no implementation of `test` or `test1` either, class `D`, being non-abstract, needs to provide some implementation of its entire interface.

**Comment by zmtzawqlp** [comment]
> Doing `with C` does not mix in any members at all, because `C` does not declare any.
> 
> The `C` mixin only `implements A`, it does not extend `A` or mix it in (it can't, it's a mixin), so it inherits no method _implementation_ from `A`. The `C` mixin is completely empty of implementation.
> 
> Since the super class `F` has no implementation of `test` or `test1` either, class `D`, being non-abstract, needs to provide some implementation of its entire interface. It inherits nothing from either `F` or `C`.

is there any way , do following code?

``` dart
class F {}

mixin A on F {
  void test();
}

mixin B on A {
  void test1();
}

class C extends F with B {}
```

**Comment by zmtzawqlp** [comment]
> Doing `with C` does not mix in any members at all, because `C` does not declare any.
> 
> The `C` mixin only `implements A`, it does not extend `A` or mix it in (it can't, it's a mixin), so it inherits no method _implementation_ from `A`. The `C` mixin is completely empty of implementation.
> 
> Since the super class `F` has no implementation of `test` or `test1` either, class `D`, being non-abstract, needs to provide some implementation of its entire interface. It inherits nothing from either `F` or `C`.

thanks for your quickly reply, if i want to print 'C' with following code, what can i do? is there any way to do this ,thanks

``` dart
void main() {
  E d= E();
  d.test();
  F f= F();
  f.test();
}


class A {
  void test() {
    print('A');
  }
}

class B extends A {
  @override
  void test() {
    print('B');
  }
}

mixin C on A {
  @override
  void test() {
    print('C');
    super.test();
  }
}

mixin D on A implements C {
  void test1() {}
}

class E extends A with D {}


class F extends B with D{}
```

**Comment by zmtzawqlp** [comment]
> Doing `with C` does not mix in any members at all, because `C` does not declare any.
> 
> The `C` mixin only `implements A`, it does not extend `A` or mix it in (it can't, it's a mixin), so it inherits no method _implementation_ from `A`. The `C` mixin is completely empty of implementation.
> 
> Since the super class `F` has no implementation of `test` or `test1` either, class `D`, being non-abstract, needs to provide some implementation of its entire interface. It inherits nothing from either `F` or `C`.

thanks for your quickly reply, if i want to print 'C' with following code, what can i do? is there any way to do this ,thanks

``` dart
void main() {
  E d= E();
  d.test();
  F f= F();
  f.test();
}


class A {
  void test() {
    print('A');
  }
}

class B extends A {
  @override
  void test() {
    print('B');
  }
}

mixin C on A {
  @override
  void test() {
    print('C');
    super.test();
  }
}

mixin D on A implements C {
  void test1() {}
}

class E extends A with D {}


class F extends B with D{}
```

**Comment by zmtzawqlp** [comment]
something like 

`mixin D with C`

**Comment by lrhn** [comment]
Dart does not have composite mixins, you can only apply one level of declarations per mixin application.
You need to write:
```dart
class F extends B with C, D {}
```
in order to get the implementations of methods from both `C` and `D`.

The only other alternative is to copy the method from `C` into `D`, which is not generally a good idea (duplicating code is extra work and harder to maintain).

**Comment by zmtzawqlp** [comment]
> The only other alternative is to copy the method from `C` into `D`, which is not generally a good idea (duplicating code is extra work and harder to maintain).

yes, i have done something like this in my code, but it's not good for me. any plan for` mixin D with C` for now?

**Comment by lrhn** [comment]
Introducing composite mixins is not on our current or near priority list, sorry.
